File: lintcode/construct_binary_tree_from_inorder_and_post_order_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param inorder : A list of integers that inorder traversal of a tree
    @param postorder : A list of integers that postorder traversal of a tree
    @return : Root of a tree
    """
    def buildTree(self, inorder, postorder):
        if inorder:
            root = TreeNode(postorder.pop())
            index = inorder.index(root.val)
            right, inorder[index:] = inorder[index + 1:], []
            post_right, postorder[index:] = postorder[index:], []
            root.left, root.right = self.buildTree(inorder, postorder), \
                                    self.buildTree(right, post_right)
            return root

File: leetcode/construct_binary_tree_from_inorder_and_postorder_traversal.js
/* Given inorder and postorder traversal of a tree, construct the binary tree.
 * Time Complexity: O(n)
 * Space Complexity: O(n), O(h) is achievable assuming partition in-place
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
    if (inorder.length > 0) {
        var root = new TreeNode(postorder.pop());
        var index = inorder.indexOf(root.val);
        var right = inorder.splice(index + 1, inorder.length - index - 1);
        inorder.pop();
        var post_right = postorder.splice(index, postorder.length - index);
        root.left = buildTree(inorder, postorder);
        root.right = buildTree(right, post_right);
        return root;
    }
    return null;
};
